### PR TITLE
[4.0] validate username

### PR DESCRIPTION
**Pull request by alikon**
Pull Request for Issue #19301 .

### Summary of Changes

porting  #23918 to 4 branch

### Testing Instructions
see #23918

### CODE REVIEW

**Comment by ladyjer** [comment]
Hi,

test almost successfully, but the INSTL_USERNAME_INVALID placeholder is missing and there is a message talking about wrong db credentials, which actually are correct.

screen shot 2019-02-26 at 21 47 01<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24024">issues.joomla.org/tracker/joomla-cms/24024</a>.</sub>

File: installation/src/Form/Rule/UsernameRule.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	public function test(\SimpleXMLElement $element, $value, $group = null, Registry $input = null, Form $form = null)
+	{
+		$filterInput = InputFilter::getInstance();
+		if (preg_match('#[<>"\'%;()&\\\\]|\\.\\./#', $value) || strlen(utf8_decode($value)) < 2
+			|| $filterInput->clean($value, 'TRIM') !== $value
+			|| strlen(utf8_decode($value)) > $element['size'])
+		{
+			return false;
+		}
+		return true;
```
Comment (Quy) [inline installation/src/Form/Rule/UsernameRule.php:49]:
```suggestion

		return true;
```

File: installation/src/Form/Rule/UsernameRule.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	 *                                       full field name would end up being "bar[foo]".
+	 * @param   Registry           $input    An optional Registry object with the entire data set to validate against the entire form.
+	 * @param   Form               $form     The form object for which the field is being tested.
+	 *
+	 * @return  boolean  True if the value is valid, false otherwise.
+	 */
+	public function test(\SimpleXMLElement $element, $value, $group = null, Registry $input = null, Form $form = null)
+	{
+		$filterInput = InputFilter::getInstance();
+		if (preg_match('#[<>"\'%;()&\\\\]|\\.\\./#', $value) || strlen(utf8_decode($value)) < 2
```
Comment (Quy) [inline installation/src/Form/Rule/UsernameRule.php:42]:
```suggestion

		if (preg_match('#[<>"\'%;()&\\\\]|\\.\\./#', $value) || strlen(utf8_decode($value)) < 2
```

**Comment by alikon** [comment]
@ladyjer that's why is still a Draft one :tongue:

**Comment by alikon** [comment]
the missing string has been added thanks to @ladyjer

File: installation/language/en-GB/en-GB.ini [hunk 3]
```diff
@@ -68,6 +68,7 @@ INSTL_DATABASE_PREFIX_MSG="The table prefix must start with a letter, optionally
 INSTL_DATABASE_TYPE_DESC="Select the database type."
 INSTL_DATABASE_USER_DESC="Either a username you created or a username provided by your host."
 INSTL_DATABASE_VALIDATION_ERROR="Check your database credentials, database type, database name or hostname. If you have MySQL 8 installed then please read this <a href=\"https://docs.joomla.org/Joomla_and_MySQL_8#Workaround_to_get_Joomla_working_with_MySQL_8\" target=\"_blank\">wiki</a> for more information."
+INSTL_USERNAME_INVALID="Joomla Super User account name invalid, avoid using special characters like '<, >, %, (, ), &' and quotes"
```
Comment (Quy) [inline installation/language/en-GB/en-GB.ini]: Here is the error message in the admin. Preferably they should be the same.
`Please enter a valid username. No space at beginning or end, at least 2 characters and must not have the following characters: < > \ " ' % ; ( ) &.`

Comment (alikon) [reply installation/language/en-GB/en-GB.ini]: done

**Comment by Quy** [comment]
The placeholder `%d` will be an issue.

**Comment by alikon** [comment]
placeholder issue fixed

File: installation/language/en-GB/en-GB.ini [hunk 4]
```diff
@@ -68,6 +68,7 @@ INSTL_DATABASE_PREFIX_MSG="The table prefix must start with a letter, optionally
 INSTL_DATABASE_TYPE_DESC="Select the database type."
 INSTL_DATABASE_USER_DESC="Either a username you created or a username provided by your host."
 INSTL_DATABASE_VALIDATION_ERROR="Check your database credentials, database type, database name or hostname. If you have MySQL 8 installed then please read this <a href=\"https://docs.joomla.org/Joomla_and_MySQL_8#Workaround_to_get_Joomla_working_with_MySQL_8\" target=\"_blank\">wiki</a> for more information."
+INSTL_USERNAME_INVALID="Please enter a valid username. No space at beginning or end, at least 30 characters and must <strong>not</strong> have the following characters: < > \ &quot; ' &#37; ; ( ) &."
```
Comment (Quy) [inline installation/language/en-GB/en-GB.ini]:
```suggestion
INSTL_USERNAME_INVALID="Please enter a valid username. No space at beginning or end, at least 2 characters and must <strong>not</strong> have the following characters: < > \ &quot; ' &#37; ; ( ) &."
```

File: installation/language/en-US/en-US.ini [hunk 5]
```diff
@@ -66,6 +66,8 @@ INSTL_DATABASE_PREFIX_DESC="Enter a table prefix or use the randomly generated o
 INSTL_DATABASE_TYPE_DESC="Select the database type."
 INSTL_DATABASE_USER_DESC="Either a username you created or a username provided by your host."
 INSTL_DATABASE_VALIDATION_ERROR="Check your database credentials, database type, database name or hostname. If you have MySQL 8 installed then please read this <a href=\"https://docs.joomla.org/Joomla_and_MySQL_8#Workaround_to_get_Joomla_working_with_MySQL_8\" target=\"_blank\">wiki</a> for more information."
+INSTL_USERNAME_INVALID="Please enter a valid username. No space at beginning or end, at least 30 characters and must <strong>not</strong> have the following characters: < > \ &quot; ' &#37; ; ( ) &."
```
Comment (Quy) [inline installation/language/en-US/en-US.ini]:
```suggestion
INSTL_USERNAME_INVALID="Please enter a valid username. No space at beginning or end, at least 2 characters and must <strong>not</strong> have the following characters: < > \ &quot; ' &#37; ; ( ) &."
```

**Comment by HLeithner** [comment]
@alikon would you like to bring this PR back?

**Comment by alikon** [comment]
reopened as requested

**Comment by roland-d** [comment]
@alikon Can you update this PR with the current 4.0-dev branch so I can get this tested? Also move this to a real PR and not a draft? Thanks.

**Comment by alikon** [comment]
i don't think this is needed anymore due to the merge of 3.9.4 in 4.0

**Pull request closed without merging**